packages/jsii-pacmak/lib/targets/python/type-name.ts

```diff
@@ -0,0 +1,354 @@
+  readonly #wrapped: TypeName;
+
+  public constructor(wrapped: TypeName) {
+    this.#wrapped = wrapped;
+  }
+
+  public pythonType(context: NamingContext) {
+    const optionalType = this.#wrapped.pythonType({ ...context, ignoreOptional: true });
+    if (context.ignoreOptional || this.#wrapped === Primitive.ANY) {
+      return optionalType;
+    }
+    return `typing.Optional[${optionalType}]`;
+  }
+
+  public requiredImports(context: NamingContext) {
+    const wrapped = this.#wrapped.requiredImports({ ...context, ignoreOptional: true });
+    if (context.ignoreOptional) {
+      return wrapped;
+    }
+    return mergePythonImports(wrapped, { typing: new Set(['']) });
+  }
+}
+
+class Primitive implements TypeName {
+  private static readonly BOOL = new Primitive('bool', {});
+  private static readonly DATE = new Primitive('datetime.datetime', { 'datetime': new Set(['']) });
+  private static readonly JSII_NUMBER = new Primitive('jsii.Number', {}); // "jsii" is always already imported!
+  private static readonly STR = new Primitive('str', {});
+  private static readonly JSON = new Primitive('typing.Mapping[typing.Any, typing.Any]', { 'typing': new Set(['']) });
+
+  public static readonly ANY = new Primitive('typing.Any', { 'typing': new Set(['']) });
+  public static readonly NONE = new Primitive('None', {});
+
+  public static of(type: PrimitiveTypeReference): TypeName {
+    switch (type.primitive) {
+      case PrimitiveType.Boolean:
+        return Primitive.BOOL;
+      case PrimitiveType.Date:
+        return Primitive.DATE;
+      case PrimitiveType.Number:
+        return Primitive.JSII_NUMBER;
+      case PrimitiveType.String:
+        return Primitive.STR;
+      case PrimitiveType.Json:
+        return Primitive.JSON;
+      case PrimitiveType.Any:
+      default:
+        return Primitive.ANY;
+    }
+  }
+
+  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+  readonly #pythonType: string;
+  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+  readonly #requiredImports: PythonImports;
+
+  private constructor(pythonType: string, requiredImports: PythonImports) {
+    this.#pythonType = pythonType;
+    this.#requiredImports = requiredImports;
+  }
+
+  public pythonType() {
+    return this.#pythonType;
+  }
+
+  public requiredImports() {
+    return this.#requiredImports;
+  }
+}
+
+class Union implements TypeName {
+  // eslint-disable-next-line @typescript-eslint/explicit-member-accessibility
+  readonly #options: readonly TypeName[];
+
+  public constructor(options: readonly TypeName[]) {
+    this.#options = options;
+  }
+
+  public pythonType(context: NamingContext) {
+    return `typing.Union[${this.#options.map(o => o.pythonType(context)).join(', ')}]`;
+  }
+
+  public requiredImports(context: NamingContext) {
+    return mergePythonImports(
+      ...this.#options.map(o => o.requiredImports(context)),
+      { typing: new Set(['']) },
```

Comment:
This seems to be always needed, any way to generalize and avoid the duplication everywhere?

Reply:
It's not always needed. It's *often* needed. I can introduce a constant for `IMPORT_TYPING`, but that would be as much as I can de-duplicate here...

Reply:
Would definitely be more readably and less error prone. You can also change this to something like:

```ts
return importsWithTyping(this.#options.map(o => o.requiredImports(context))
```

Or even:

```ts
return importsWithTyping(this.#options);
```